### ISSUE TITLE
Questions about html escape

### ISSUE THREAD

**Issue by dstrelau** [slackhq/hubot-slack#43]
Hi folks. I'm getting unexpected behavior when trying to send formatted links.

I'm attempting to send a response from hubot with a link label, like so:

```
msg.send "<https://github.com/link/to/a/PR|myrepo #42> fix some broken"
```

What I'd expect to see in Slack is this, with `{{ … }}` representing hyperlinked text with the URL I passed.

```
{{myrepo #42}} fix some broken
```

Instead, I see this, with literal brackets and the link in the wrong place:

```
<{{myrepo}} #42> fix some broken
```

Looking around, it's pretty evident that this is due to [the `escapeHTML` call](https://github.com/tinyspeck/hubot-slack/blob/master/src/slack.coffee#L28) that occurs before the message is sent. In fact, [removing that call](https://github.com/dstrelau/hubot-slack/commit/3ff476e59ca01a6634a7160d740ad837f2f98528) gives me the result I'd expect, with the properly parsed link.

I'm trying to understand how https://api.slack.com/docs/formatting works in the context of hubot, but I'm a bit confused. What is the best way to allow for what I want to do without screwing up everyone else's hubot? Should I create `msg.sendRaw`? Is it actually safe to remove `escapeHTML` since slack handles most of it server-side?

**Comment by dstrelau** [comment]
@paulhammond Could you provide an update here? I'd be happy to work on the adapter code here if the API fixes you mentioned have taken place.

**Comment by lukerollans** [comment]
Also experiencing this issue! When our hubot sends pull request related notices to Slack, the message should appear as `repo#pr_number`, eg `my-repo#35`. This means that the msg is sent to Slack like ...

`https://github.com/organization/repo/pull/25|repo#25', however the pipe character is being url encoded so the entire message appears broken.

Watching this issue closely!

**Comment by technicalpickles** [comment]
I was able to reproduce this pretty easily. I'm not sure why it's being escaped yet, as there doesn't to seem to be any obvious escaping code in the adapter. It might be in the node-slack, or maybe it's a parameter to the API that needs to be specified to parse it as mrkdwn.

I found a workaround though. It's not obvious, but you can use slack attachments to do this:

```coffeescript
robot.emit 'slack.attachment', {text: "<https://github.com/link/to/a/PR|myrepo #42> fix some broken">}
```

**Comment by technicalpickles** [comment]
I did find a difference between `adapter.send` and `adapter.customMessage` (which is used by slack.attachment) event: `send` uses [channel.send](https://github.com/slackhq/node-slack-client/blob/master/src/channel.coffee#L85-L87) which calls [channel.sendMessage](https://github.com/slackhq/node-slack-client/blob/master/src/channel.coffee#L102-L104) which calls [client._send](https://github.com/slackhq/node-slack-client/blob/master/src/client.coffee#L524-L532),  compared to `customMessage` which calls [channel.postMessage](https://github.com/slackhq/node-slack-client/blob/master/src/channel.coffee#L89-L97) which [calls the API](https://github.com/slackhq/node-slack-client/blob/master/src/client.coffee#L534-L565).

So the difference comes down to `adapter.send` uses the RTM to send messages back, where attachments use the web API. Over on https://api.slack.com/rtm#sending_messages it says:

> The RTM API only supports posting simple messages formatted using our default message formatting mode. It does not support attachments or other message formatting modes. To post a more complex message as a user clients can call the chat.postMessage Web API method with as_user set to true.

So it seems like this should work as is, which makes me think there is a bug in the RTM. I'm not sure if there's any advantage to using RTM for responses, or if we can always just use the web API instead.

**Comment by technicalpickles** [comment]
I got confirmation from Slack that the RTM API doesn't support link formatting:

> Your conclusion is correct, the RTM API currently doesn't support link formatting. Using the chat.postMessage method is what we would recommend in cases where you need more than just bare bones formatting, and you shouldn't notice any downsides to using it instead of the RTM API.

So, sounds like we update to always use the chat.postMessage API.

**Comment by DEGoodmanWilson** [comment]
Closing this as fixed in 4.0. Please reopen if I am mistaken!
